File: Trees/0543. Diameter of Binary Tree/0543.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int result = 0;
        DFS(root, result);
        return result;

    }
private: 
    int DFS(TreeNode* root, int& result) {
        if (root == NULL) {
            return 0;
        }

        int left = DFS(root->left, result);
        int right = DFS(root->right, result);

        result = max(result, left + right);
        return 1 + max(left, right);
    }
};

File: Trees/0543. Diameter of Binary Tree/0543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # diameter is basically the heights of l + r, not the height of the current node which would be
        # l + r + 1. For this reason, we need to keep track of a height and a diameter, and since the
        # max diameter could be anywhere in the tree, we need to max() it every time. 

        diameter = 0

        # you can't directly pass something by reference in python, so the easiest method of keeping
        # track of a diameter is to make the function inside of the solution function
        def dfs(root):
            # to say that it is a global function
            nonlocal diameter

